### ISSUE TITLE
shed_upload problems

### ISSUE THREAD

**Issue by jmchilton** [galaxyproject/planemo#131]
Building a new abstraction and test cases for repository creation and I noticed some problems with upload.

 - [ ] ``.shed.yml`` ignored broken.
 - [ ] ``.git`` directories are not being properly excluded from repository creation. 

I'll push a fix for these shortly along with tests.

**Comment by erasche** [comment]
> best practice should probably be one tool shed repository per tool and this is not how many tools are stored

eh, really? Just use more meta packages (suite/tool_collections) to define collections of tools? That's always felt like a LOT of overhead (see devteam tools, https://github.com/galaxyproject/tools-devteam/tree/master/tool_collections/gatk, so many folders/test-data copies/.shed.yml files to hold tools that 100% of the time ship/install together).

**Comment by martenson** [comment]
@erasche it has reproducibility, updating and other benefits to keep tool repositories atomic.
Current implementation of suites needs some functionality to make the metapackages more convenient but generally 1 tool per repo is the preferred way to go. I would call it best practice.

**Comment by erasche** [comment]
@martenson huh. Yuck. Sorry, it'll take me some time to be convinced of this idea. 

I can understand it being best practice on TS designer viewpoint, but it feels anti-developer, as we like collecting all of the tools + reusable macros in a single folder. If you want 1 tool:1 folder on disk, that sort of mapping will definitely be needed in planemo, I don't want to move from maintaining a single folder of 10 tools, to 10 folders of tools with a macros.xml that's *above* them, rather than next to the tools.

As for updating? It'd mean if we push a tool update, then we need to re-push the metapackage, rather than just pushing the collection of tools. Again, not exactly developer friendly. (It also means I'll have to implement #71.) I recognise that planemo is here to help with this transition, and if it's the most technically correct way to do it in the TS, then that's the way to  go, it's just a bit of a change for me.

**Comment by jmchilton** [comment]
@erasche  That is exactly what I am trying to get away from needing with work like e8d2e227583cb548d9c7c70c89163051066849fa and #26. Ideally you will just have a flat list of tool files in the source folder and some flag in ``.shed.yml`` to de-multiplex them into separate repositories during upload, lint, etc....

The point that it is tough on deployers is also well taken - but @martenson is redoing search and hopefully that will give a lot more weight to suites and this problem will also be minimized.

Back to @martenson original comment - he said "reproducibility" but I am not sure that is really the problem here. The problem with the big multi-tool repositories is that they are bad for users (given the current implementation) - if any of the tools needs a fix then all of the tools belong to a new install-able revision. This makes in inconvenient to use any of the tools with features such as workflows or re-runs if just one of them is updated.

All that said - it probably is a "best practice" to use a separate repository per tool but no one is doing it and it is not reflected in the IUC recommendations so I wouldn't feel any pressure to do things this way. Once search and planemo make it really easy - hopefully the benefits of doing it that way will become clear and then it will become a true community accepted best practice. 

tl;dr - We are either inconvenience the deployers and developers or the users. The hope is to do the former but to use planemo and tool shed search revamp to minimize the inconvenience.

**Comment by erasche** [comment]
@jmchilton thanks for the explanation, that was needed. 

> if any of the tools needs a fix then all of the tools belong to a new install-able revision

yeah, I've experienced this myself, the question of "oh, did something *actually* change in this tool, or was it in one of the related ones?"

If planemo is going to handle all the de-multiplexing, then flipping a single flag is definitely do-able. If you decide on a flag name we'll add it to the docs and be happy to be the first to test it out (at scale)!

**Comment by bgruening** [comment]
I agree here with @martenson and @jmchilton. In a ideal world we have one repository per tool. But before planemo this was simply to much overhead for a developer.
Currently, I'm thinking more in the direction of handling the tool splitting into multiple repositories as part of the TS.  For example, the TS should have all informations needed to split BEDtools into 20 separated repositories. If information is needed we can guide the TS via `.shed.yml`.

**Comment by erasche** [comment]
@bgruening yeah, either in TS/planemo, as long as it isn't on the part of the developer, because that's just too much work.
